pkg/registry/scheduling/rest/storage_scheduling.go

```diff
@@ -49,6 +63,64 @@ func (p RESTStorageProvider) v1alpha1Storage(apiResourceConfigSource serverstora
 	return storage
 }
 
+func (p RESTStorageProvider) PostStartHook() (string, genericapiserver.PostStartHookFunc, error) {
+	return PostStartHookName, AddSystemPriorityClasses(), nil
```

Comment:
can i delete them?  what happens if i do? will they come back only on restart?

Reply:
If they are deleted, no more new Pods with these priority classes can be created. These PriorityClasses are added back on API server restart.

Reply:
I think it also means no new critical pods could be created. Can we avoid that? Say something like Delete would return error, if priorityclass name is one of systemclustercritical or systemnodecritical? Not sure if that takes flexibility away from user.

Reply:
Looks like we don't have a validation mechanism in our REST layer for delete operations. So, I am not sure if it is possible to prevent deletion of these priority classes.

Reply:
Thanks for the information but I did a quicksearch in the codebase and it seems there are few things like `immortalNamespaces`(basically 'kube-system' and 'defaut' etc), where at the admission plugin level we are prohibiting user to delete them.(https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/admission/plugin/namespace/lifecycle/admission.go#L77). Wondering if we could do the same thing but we need think it can be a different admission controller from priority AdmissionController.

Reply:
> Looks like we don't have a validation mechanism in our REST layer for delete operations. So, I am not sure if it is possible to prevent deletion of these priority classes.

You can prevent it in the rest handler's Delete function

Reply:
@liggitt You mean somewhere here? - https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/scheduling/priorityclass/registry.go#L81

Reply:
inside the impl of `Delete` (which the priorityclassstorage would need to override, put the guard in, then delegate, like:

```go
func (r *REST) Delete(ctx genericapirequest.Context, name string, options *metav1.DeleteOptions) (runtime.Object, bool, error) {
	// if name is in set of expected system priority classes, forbid and return
	return r.store.Delete(ctx, name, options)
}
```

Reply:
Understood. Thanks @liggitt

Reply:
Thanks, @liggitt  and @ravisantoshgudimetla. I added the check.

Reply:
thanks for updating.